### ISSUE TITLE
Broken completion for pacman and other wrappers

### ISSUE THREAD

**Issue by ariasuni** [fish-shell/fish-shell#4763]
I use fish 2.7.1 and when I try to complete options for pacman (without automatically generated completions):

- To begin with, this is seriously wrong:

```
> pacman -Qddddddd<TAB>
-Qdddddddb                         (Alternative database location)
-Qdddddddo                     (Cherche les paquet contenant le motif)
[…]
```

- Translation is broken. The description is sometimes in French but when it is, this is the description for the wrong command.

For example, `pacman -S` gives me errors like:
> -Qt  (Afficher seulement les paquets pouvant être mis à niveau)
> -Qu  (Afficher seulement les paquets pouvant être mis à niveau)

which is obviously false, or:

> -Qd   (Installer/supprimer les dépendances de construction)

when `pacman --help` give me:

> -d, --deps               liste tous les paquets installés en tant que dépendances [filtre]

I don’t understand where the translation comes from since pacman is not distributed with a French man page, and the `pacman --help` command shows a different string.

For example, when I try to complete `pacman -S` I get:
> -Sc    (Supprimer tous les fichiers gz de la cache)

and `pacman -S --help` give me:
> -c, --clean              supprime les paquets obsolètes du cache (-cc pour tous)

**Comment by faho** [comment]
>Translation is broken. The description is sometimes in French but when it is, this is the description for the wrong command.

Yup. The issue here is that gettext tries to find translations on its own by doing fuzzy matching. These need to be manually sifted through.

And as luck would have it, @PenegalECI and @lourkeur have done a bunch of work on the french translations for 3.0. I'm assuming this is in there.

>since pacman is not distributed with a French man page

Those are _our_ translations, not pacmans.

>To begin with, this is seriously wrong:

What exactly do you mean here? That `pacman -Qdddd...` still completes options? That is actually something that pacman accepts. You can add "-d" as often as you like. Now, it'd be nicer if it was just once for "-d", but it accepts e.g. "-t" twice (the second makes it also ignore optional dependencies).

**Comment by ariasuni** [comment]
> And as luck would have it, @PenegalECI and @lourkeur have done a bunch of work on the french translations for 3.0. I'm assuming this is in there.

Nice! Shouldn’t you use pacman’s own translations then? Mostly by curiosity, where are the translations and where is the code handling it?

> What exactly do you mean here? That pacman -Qdddd... still completes options? That is actually something that pacman accepts. You can add "-d" as often as you like. Now, it'd be nicer if it was just once for "-d", but it accepts e.g. "-t" twice (the second makes it also ignore optional dependencies).

The man page document which option can be accepted twice. If it is possible, I would like the following behavior for completion:

- `pacman -Qd` doesn’t display `-Qdd` option
- `pacman -Qt` displays`-Qtt` and describe what happens when the flag is used twice
- `pacman -Qtt` doesn’t display `-Qttt` option

I also updated the first comment with another grief I have about the fish pacman completion.

I don’t have any prior knowledge of fish completion nor much time; but if it’s easy enough and given enough indications, I could try to improve fish completion, at least a bit. I use Arch Linux at home and at work, and this is really annoying sometimes. I want to make the switch from bash/zsh to fish painless for Arch Linux users.

**Comment by faho** [comment]
>Shouldn’t you use pacman’s own translations then?

No, because they aren't written for our purpose. That means they might be too long () or technical. Space is at a premium here, because if we manage to get it under a certain length we can use more columns. And automated truncation just doesn't work all that well.

We can use them to write translations (and descriptions in general), but we shouldn't just use them in an automated way.

What we should do is remove the fuzzy translations, at least by default. I'll try to figure out how to do that. We can use them as a starting point, but it's probably better to show someone an accurate english version rather than a misleading native one.

At a glance, this seems as simple as removing "--use-fuzzy" from our msgfmt invocation.

>Mostly by curiosity, where are the translations and where is the code handling it?

The code handling it is in src/complete.cpp (it tries to translate every description given in a completion script - look for `C_(`). The translations for the entirety of fish, including completions, are in po/$LANG.po, e.g. [po/fr.po](https://github.com/fish-shell/fish-shell/blob/master/po/fr.po).

>The man page document which option can be accepted twice. If it is possible, I would like the following behavior for completion:

It would be possible, but this is a general problem, and adding it to every single command does not scale. Even for a single command, it's a PITA. I just checked the pacman completion script, and adding this would require touching ~90% of the lines.

And considering that the workaround is quite simple - just don't select the option again - and the failure mode is so benign - pacman'll still accept it, it's not really a priority.

The general idea has been discussed in #1315.

**Comment by ariasuni** [comment]
The first point is not going to be fixed because it’s too complex/not worth it, if I understand correctly. That seems fine in my opinion.

For the second point, @faho said that «What we should do is remove the fuzzy translations, at least by default.», so maybe we should leave it open until it’s fixed.

For the third point, I’m not sure what’s possible but that would be a huge improvement on the current situation. It concerns really few commands, for example `pacman -Qo`, but it’s a pain to use right now.

**Comment by faho** [comment]
>However, I didn’t manage to make fish completes files after -Qo with the same tricks.

@ariasuni: That's probably because the completion explicitly inhibits filenames via `complete -f` or `-x`. Remove that and it should work.

**Comment by ariasuni** [comment]
I tried to do that, but then other commands don’t complete for package names. I think I exhausted obvious solutions, I took one or two hours to look at the options of `complete` and some other completions files, but nothing worked out.

**Comment by zanchey** [comment]
Fuzzy completions have been removed in 14d59c53f63075aa405551533ae9ac08dc9e25a9 and #4847, so we'll leave this open for the completion improvements.

**Comment by esclerofilo** [comment]
Is there an explanation for the current behavior, where `pacman -Qo` doesn't have correct completions but `pacman -Q -o` and `pacman -Q --owns` do?
